app/models/school.rb

```diff
@@ -64,6 +64,20 @@ class School < ApplicationRecord
     joins(:school_cohorts).where(school_cohorts: { cohort_id: cohort.id, opt_out_of_updates: true })
   }
 
+  scope :all_ecf_participants_validated, lambda {
+    left_outer_joins(ecf_participant_profiles: %i[ecf_participant_eligibility ecf_participant_validation_data])
+      .group(:id)
+      .having(<<~SQL)
+        SUM(
+          CASE
+            WHEN COALESCE(ecf_participant_eligibilities.id, ecf_participant_validation_data.id) IS NULL
+            THEN 1
+            ELSE 0
+          END
+          ) = 0
+    SQL
+  }
```

Comment:
Not a big fan of this scope.

One problem with it is 'what about withdrawn participants?'. Which is luckily remedied by using `active_ecf_participant_profiles` in the outer joins.

Should we maybe add `not_validated_ecf_participant_profiles` to SchoolCohort, and link them using a `through` association here? That would let us do something like `scope :all_ecf_participants_validated, -> { where.missing(:not_validated_ecf_participant_profiles) }`

Thoughts?

Reply:
yeah I'm not super proud of it, I was going to put a comment asking for any other ideas than a big splat of raw SQL. I'll give your idea whirl

Reply:
Unfortunately it just... doesn't work. Since this is due for tomorrow, I don't think there is time to figure out the problem:
```
[1] pry(#<RSpec::ExampleGroups::School::Scopes::AllEcfParticipantsValidated>)> School.all_ecf_participants_valid
ated
=> []
[2] pry(#<RSpec::ExampleGroups::School::Scopes::AllEcfParticipantsValidated>)> School.joins(:not_validated_ecf_p
articipant_profiles)
=> [#<School:0x000000012100d468
  id: "750e9893-75a3-4ec5-8c9d-3852a3d31726",
  created_at: Thu, 30 Sep 2021 14:54:20.369030000 UTC +00:00,
  updated_at: Thu, 30 Sep 2021 14:54:20.369030000 UTC +00:00,
  urn: "0519848",
  name: "White Academy",
  school_type_code: 1,
  address_line1: "982 Milagro Meadow",
  address_line2: nil,
  address_line3: nil,
  postcode: "72425-2844",
  network_id: nil,
  domains: ["zboncak.net"],
  school_type_name: nil,
  ukprn: nil,
  school_phase_type: nil,
  school_phase_name: nil,
  school_website: nil,
  school_status_code: 1,
  school_status_name: nil,
  secondary_contact_email: nil,
  primary_contact_email: "johnetta@collier.co",
  administrative_district_code: "E123",
  administrative_district_name: nil,
  slug: "0519848-white-academy",
  section_41_approved: false>,
 #<School:0x000000012100d3a0
  id: "1647ad37-04b0-4b05-a494-9fbf58aeea41",
  created_at: Thu, 30 Sep 2021 14:54:20.563396000 UTC +00:00,
  updated_at: Thu, 30 Sep 2021 14:54:20.563396000 UTC +00:00,
  urn: "0610139",
[3] pry(#<RSpec::ExampleGroups::School::Scopes::AllEcfParticipantsValidated>)> School.find("750e9893-75a3-4ec5-8
c9d-3852a3d31726").not_validated_ecf_participant_profiles
=> []
```

Code added:
In school cohort:
```
  has_many :not_validated_ecf_participant_profiles, -> {
    ecf.active_record.where.missing(:ecf_participant_eligibility, :ecf_participant_validation_data)
  }, class_name: "ParticipantProfile::ECF"
```

In school:
```
  has_many :not_validated_ecf_participant_profiles, through: :school_cohorts
  
    scope :all_ecf_participants_validated, lambda {
      where.missing(:not_validated_ecf_participant_profiles)
    }

```